physicalplan/function/function.go

```diff
@@ -1,247 +1,479 @@
+		return nil, err
+	}
+
+	if len(input) == 0 {
+		return nil, nil
+	}
+
+	scalarInput := make([][]model.StepVector, len(o.scalarArgs))
+	if len(o.scalarArgs) != 0 {
+		for i := range o.scalarArgs {
+			scalarInput[i], err = o.scalarArgs[i].Next(ctx)
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	var scalarPoints []float64
+	if len(o.funcExpr.Func.ArgTypes) > 1 && len(scalarInput) > 0 {
+		for j := range scalarInput {
+			scalarPoints = append(scalarPoints, scalarInput[j][0].Samples[0])
+		}
+	}
+
+	out := o.next.GetPool().GetVectorBatch()
+	for _, vector := range input {
+		step := o.next.GetPool().GetStepVector(vector.T)
+		for i := range vector.Samples {
+			// Call function by separately passing major input and scalar constants.
+			result := o.call(FunctionArgs{
+				Labels:       o.series[0],
+				Points:       []promql.Point{{V: vector.Samples[i]}},
+				StepTime:     step.T,
+				ScalarPoints: scalarPoints,
+			})
+
+			step.Samples = append(step.Samples, result.V)
+			step.SampleIDs = append(step.SampleIDs, vector.SampleIDs[i])
+		}
+
+		out = append(out, step)
+		o.next.GetPool().PutStepVector(vector)
+	}
+	o.next.GetPool().PutVectors(input)
+
+	return out, nil
+}
+
+// Process single arg range vector input functions.
+func (o *functionSelector) rangeVectorNext(ctx context.Context) ([]model.StepVector, error) {
+	input, err := o.next.Next(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(input) == 0 {
+		return nil, nil
+	}
+
+	// We inject function call in tree, so only need to relay input from here.
+	return input, nil
+}
+
+// Handle unimplemented cases.
+func (o *functionSelector) defaultNext(ctx context.Context) ([]model.StepVector, error) {
+	return nil, errors.Wrapf(parse.ErrNotImplemented, "got: %s", o.funcExpr)
+}
+
+func (o *functionSelector) loadSeries(ctx context.Context) error {
+	var err error
+	o.once.Do(func() {
+		if o.funcExpr.Func.Name == "vector" {
+			o.series = []labels.Labels{labels.New()}
+			o.next.GetPool().SetStepSize(len(o.series))
+			return
+		}
+
+		series, loadErr := o.next.Series(ctx)
+		if loadErr != nil {
+			err = loadErr
+		}
+
+		o.series = make([]labels.Labels, len(series))
+		for i, s := range series {
+			lbls := s
+			if o.funcExpr.Func.Name != "last_over_time" {
+				lbls = dropMetricName(s)
+			}
+
+			sort.Sort(lbls)
+			o.series[i] = lbls
+		}
+
+		o.next.GetPool().SetStepSize(len(o.series))
+	})
+
+	return err
+}
+
 var InvalidSample = promql.Sample{Point: promql.Point{T: -1, V: 0}}
 
+type FunctionArgs struct {
+	Labels       labels.Labels
+	Points       []promql.Point
+	StepTime     int64
+	SelectRange  int64
+	ScalarPoints []float64
+}
+
 // FunctionCall represents functions as defined in https://prometheus.io/docs/prometheus/latest/querying/functions/
-type FunctionCall func(labels labels.Labels, points []promql.Point, stepTime int64, selectRange int64) promql.Sample
+type FunctionCall func(f FunctionArgs) promql.Sample
 
 var Funcs = map[string]FunctionCall{
```

Comment:
This map can probably be split into different categories of function, so that the arguments align with PromQL. This would make processing them easier.